server/authorization.go

```diff
@@ -345,139 +396,144 @@ func logHeaderValueOptions(okResponse *authv3.OkHttpResponse) string {
 }
 
 func (a *AuthorizationServer) notFound(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
-	tracker *prometheusRequestMetricTracker) *authv3.CheckResponse {
-	return a.createDenyResponse(req, envRequest, tracker, nil, "", rpc.NOT_FOUND)
+	tracker *prometheusRequestMetricTracker, api string) *authv3.CheckResponse {
+	return a.createConditionalEnvoyDenied(req, envRequest, tracker, nil, api, rpc.NOT_FOUND)
 }
 
 func (a *AuthorizationServer) unauthenticated(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
-	tracker *prometheusRequestMetricTracker) *authv3.CheckResponse {
-	return a.createDenyResponse(req, envRequest, tracker, nil, "", rpc.UNAUTHENTICATED)
+	tracker *prometheusRequestMetricTracker, api string) *authv3.CheckResponse {
+	return a.createConditionalEnvoyDenied(req, envRequest, tracker, nil, "", rpc.UNAUTHENTICATED)
 }
 
 func (a *AuthorizationServer) internalError(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
 	tracker *prometheusRequestMetricTracker, err error) *authv3.CheckResponse {
 	log.Errorf("sending internal error: %v", err)
-	return a.createDenyResponse(req, envRequest, tracker, nil, "", rpc.INTERNAL)
+	return a.createConditionalEnvoyDenied(req, envRequest, tracker, nil, "", rpc.INTERNAL)
 }
 
 func (a *AuthorizationServer) denied(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
 	tracker *prometheusRequestMetricTracker, authContext *auth.Context, api string) *authv3.CheckResponse {
-	return a.createDenyResponse(req, envRequest, tracker, authContext, api, rpc.PERMISSION_DENIED)
+	return a.createConditionalEnvoyDenied(req, envRequest, tracker, authContext, api, rpc.PERMISSION_DENIED)
 }
 
 func (a *AuthorizationServer) quotaExceeded(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
 	tracker *prometheusRequestMetricTracker, authContext *auth.Context, api string) *authv3.CheckResponse {
-	return a.createDenyResponse(req, envRequest, tracker, authContext, api, rpc.RESOURCE_EXHAUSTED)
+	return a.createConditionalEnvoyDenied(req, envRequest, tracker, authContext, api, rpc.RESOURCE_EXHAUSTED)
 }
 
-func (a *AuthorizationServer) createDenyResponse(req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
-	tracker *prometheusRequestMetricTracker, authContext *auth.Context, api string, code rpc.Code) *authv3.CheckResponse {
+// creates a deny (direct) response if authorization has failed unless
+// handler.allowUnauthorized is true, in which case the request will be
+// allowed to continue
+func (a *AuthorizationServer) createConditionalEnvoyDenied(
+	req *authv3.CheckRequest, envRequest *config.EnvironmentSpecRequest,
+	tracker *prometheusRequestMetricTracker, authContext *auth.Context,
+	api string, code rpc.Code) *authv3.CheckResponse {
 
-	// use intended code, not OK
+	statusCode := typev3.StatusCode_Forbidden
```

Comment:
Why not mutate tracker.statusCode directly so that createEnvoyDenied can take one fewer argument?

Reply:
I found it to be less intuitive to require callers to set multiple status codes and this way it can be done in one place.